### PR TITLE
:sparkles: adding reflection client to determine what provider provides

### LINKED ISSUE #573: [BUG] codeSnips for java.dependency rules do not work with external java provider
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Konveyor version

0.4

### Priority

Critical

### Current Behavior

_No response_

### Expected Behavior

_No response_

### How Reproducible

Always (Default)

### Steps To Reproduce

_No response_

### Environment

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by shawn-hurley**
Fixes #573 
DRAFT FOR NOW

initial implementation some TODO's left want to see what CI gives me

### CODE REVIEW

**Comment by shawn-hurley** [comment]
This is breaking because we are not updating the external providers to use the updated code in analyzer-lsp during the build here. 

going to try and fix this

File: external-providers/java-external-provider/Dockerfile [hunk 1]
```diff
@@ -1,15 +1,18 @@
 # Build the manager binary
 FROM golang:1.21 as builder
```
Comment (shawn-hurley) [inline external-providers/java-external-provider/Dockerfile:2]: Need to downgrade to 1.20

**Comment by shawn-hurley** [comment]
Hey @pranavgaikwad @eemcmullan I am really sorry this feature PR turned into a catch all PR fixing CI, making all the docker files build with same go as well as the feature. 

if you really want, I can make more commits to make it easier to review but let me know what you think

**Review by pranavgaikwad — COMMENTED** [review]
Looking good overall.

File: external-providers/generic-external-provider/Dockerfile [hunk 2]
```diff
@@ -25,6 +28,6 @@ RUN npm install -g typescript-language-server typescript
 RUN go install golang.org/x/tools/gopls@latest
 
 COPY --from=go-builder /generic-external-provider/generic-external-provider /usr/local/bin/generic-external-provider
-COPY --from=go-dep-provider /usr/local/bin/go-dependency-provider /usr/local/bin/go-dependency-provider
+COPY --from=go-dep-provider /usr/local/bin/golang-dependency-provider /usr/local/bin/golang-dependency-provider
```
Comment (eemcmullan) [inline external-providers/generic-external-provider/Dockerfile:31]: Note: will need to update this in kantra now. I can do so. I just don't want to forget

File: provider/grpc/provider.go [hunk 3]
```diff
@@ -31,16 +32,116 @@ type grpcProvider struct {
...
+			},
+		}, nil
+
+	} else {
+		// just create grpcProvider
+		return &gp, nil
+	}
+}
+
+func checkServicesRunning(refClt *reflectClient.Client, log logr.Logger) ([]string, error) {
```
Comment (eemcmullan) [inline provider/grpc/provider.go:129]: +100

File: external-providers/yq-external-provider/Dockerfile [hunk 4]
```diff
@@ -1,13 +1,15 @@
 FROM golang:1.20 as go-builder
 
+copy / /analyzer-lsp
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply external-providers/yq-external-provider/Dockerfile:3]: I don't think so, I think that is moving the `/` context into `/analyzer-lsp` in the image.

File: .github/workflows/demo-testing.yml [hunk 5]
```diff
@@ -38,62 +38,70 @@ jobs:
...
-          build_image golang-dependency-provider golang-dependency-provider
-          build_image java-external-provider java-external-provider
-          build_image yq-external-provider yq-external-provider
-          build_image generic-external-provider generic-external-provider
-          popd
-          build_image . analyzer-lsp
-          
+          make build-external
+          podman build -t quay.io/konveyor/analyzer-lsp:latest -f Dockerfile .
+        # run: |
```
[the remark this answers is not in the document]
Comment (shawn-hurley) [reply .github/workflows/demo-testing.yml]: updated

**Pull request merged**